File: solutions/haskell/hamming/1/src/Hamming.hs
module Hamming (distance) where

distance :: String -> String -> Maybe Int
distance [] [] = Just 0
distance [] (_ : _) = Nothing
distance (_ : _) [] = Nothing
distance (x : xs) (y : ys) = case (distance xs ys) of
    (Just n) -> Just (thisCount + n)
    Nothing  -> Nothing
  where
    thisCount = if x == y then 0 else 1

File: solutions/erlang/hamming/1/src/hamming.erl
-module(hamming).
-export([distance/2]).

distance(XS,YS) when length(XS) /= length(YS) -> {error, badarg};
distance([], []) -> 0;
distance([X|XS], [Y|YS]) when X /= Y -> 1 + distance(XS,YS);
distance([X|XS], [Y|YS]) when X == Y -> distance(XS,YS).
